feat(script): Deploy kubeflow to IBM Cloud IKS vpc-gen2, using this script.

ScrapCodes opened the pull request:
**Which issue is resolved by this Pull Request:** 
Resolves #

**Description of your changes:**
Added a script to create a vpc-gen2 cluster and deploy kubeflow on IBM Cloud IKS.

Tomcli commented:
Thanks @ScrapCodes, can you add the usage in the comment section of this script? also it would be great if you can also add a small readme on what this script can or cannot do.

ScrapCodes commented:
@Tomcli A readme is embedded in the PR description and script itself. Do you have a suggestion, where shall we place the read-me and the scripts?

Tomcli commented:
> @Tomcli A readme is embedded in the PR description and script itself. Do you have a suggestion, where shall we place the read-me and the scripts?

Thanks for adding the comments. Can we also copy these comments into a `README.md` and put them in the same directory? thanks.

@yhwang If you have time, can you test out this script? The goal here is to automate the experience on creating the vpc cluster for Kubeflow. Hopefully we can also use it as part of advocate workshop/session.

yhwang commented:
@Tomcli sure! I can verify it.
@ScrapCodes and @Tomcli  FYI, I also worked on a schematics project that is used to create VPC Gen2 cluster and deploy the multi-user kubeflow on it. Because of multi-user, I also integrate kubeflow with AppID service. Here is my project: https://github.com/yhwang/auto-kubeflow/.

Tomcli commented:
> @Tomcli sure! I can verify it.
> @ScrapCodes and @Tomcli FYI, I also worked on a schematics project that is used to create VPC Gen2 cluster and deploy the multi-user kubeflow on it. Because of multi-user, I also integrate kubeflow with AppID service. Here is my project: https://github.com/yhwang/auto-kubeflow/.

Thanks @yhwang, the goal of this script is to have an easier time for user to deploy VPC using their local machine as client and hopefully can be reuse with IBM toolchain. But since you already have the IBM schematics for the UI experience, we can let this script focuses on more command line driven users.

yhwang commented:
@Tomcli 
> we can let this script focuses on more command line driven users.

totally agree. However, if users want to run schematics locally, it's also doable. They just need to install terraform and ansible locally. But again, you're right. Using schematics service in ibm cloud is a better option for end users.

yhwang commented:
@ScrapCodes one suggestion regarding shell scripting: it would be good to use `shellcheck` to verify the 2 shell script files: https://github.com/koalaman/shellcheck

yhwang commented on scripts/deploy/iks/deploy-ibm-vpc.sh:
```diff
@@ -0,0 +1,405 @@
+    export WORKER_NODE_FLAVOR=${WORKER_NODE_FLAVOR}
+    END
+    )
+    echo $ENV_VARS
```
need double quote for $ENV_VARS or the line break is gone.

yhwang commented on scripts/deploy/iks/deploy-ibm-vpc.sh:
```diff
@@ -0,0 +1,405 @@
+    export CLUSTER_ZONE=${CLUSTER_ZONE}
+    export CLUSTER_NAME=${CLUSTER_NAME}
+    export WORKER_NODE_FLAVOR=${WORKER_NODE_FLAVOR}
+    END
```
should remove leading space

yhwang commented on scripts/deploy/iks/deploy-kfp-ibm-vpc.sh:
```diff
@@ -0,0 +1,134 @@
+  fi
+  local val1 = $1
+  local val2 = $2
+  local msg = $3
```
these 3 lines: no space around `=`.

yhwang commented:
the public gateway detection seems buggy, it says a public gateway already attach to the subnet, but there is no public gateway. and it proceeds to deploy the kubeflow. I have to manually turn on the public gateway.

And the `current_env` doesn't work properly so `record_current_env` can't save configurations.

after I fixed the `current_env` function. I tried `./deploy-ibm-vpc.sh`. I can create cluster and deploy kubeflow successfully. but `./deploy-ibm-vpc.sh --delete-cluster="full"` failed to clean up all resources, because vpc, subnet deletion need to wait for cluster deletion. They can't be deleted while a cluster is using them.

ScrapCodes commented:
@yhwang I have tried to address the issues raised by you, and also simplified the script a bit. The way configuration file was read is also changed, from one global configuration file to one per VPC and Cluster.

Please take another look !

yhwang commented:
verified the script again today with following steps:
- `./deploy-ibm-vpc.sh --cluster-name="yhwang-script-pr" --vpc-name="yhwang-script-pr-vpc"
   I can access the kubeflow after the command finish
- `./deploy-ibm-vpc.sh --delete-cluster="full" --config-file=yhwang-script-pr-vpc-yhwang-script-pr-cluster-config.sh`
   it showed me: `Variable `VPC_NAME` is not specified, please provide `--vpc-name=value` as CLI option.`
- try to source the config file: `source yhwang-script-pr-vpc-yhwang-script-pr-cluster-config.sh` and run the delete` command again
  it showed me: 
  ```
  deleted cluster yhwang-script-pr/c1nj8nqd0i1sb9405r00 from /Users/yhwang/.kube/config
  warning: this removed your active context, use "kubectl config use-context" to select a different one
  deleted context yhwang-script-pr/c1nj8nqd0i1sb9405r00 from /Users/yhwang/.kube/config
  Failed to delete subnet 0737-3ca7f85d-5bc7-4198-8749-d075a46074f3.
  Response HTTP Status Code: 409
  Error code: conflict_field
  Error message: Vnic: 0737-1df0c454-94fc-48f3-b0ea-d522462d8942-j8zx5 is attached to Subnet: 0737-3ca7f85d-5bc7-4198-8749-d075a46074f3, please detach all network interfaces from subnet before deleting it
  Error target name: id, type: parameter
  Error target value: 0737-3ca7f85d-5bc7-4198-8749-d075a46074f3
  Trace ID: bd0235d4-9e83-47f2-9eed-d11751c8cc62
  
  Failed to delete public gateway r006-4d6d4828-b95b-471b-85f8-a84147e596fb.
  Response HTTP Status Code: 409
  Error code: conflict_field
  Error message: The Public Gateway is used by a subnet.
  Error target name: id, type: parameter
  Error target value: r006-4d6d4828-b95b-471b-85f8-a84147e596fb
  Trace ID: a491c8df-1d4d-4c2c-b925-1f8cf7bb2057
  
  Failed to delete vpc r006-17e4a0d2-81bc-407b-a74d-00a53fa314bd.
  Response HTTP Status Code: 409
  Error code: vpc_in_use
  Error message: The VPC is in use and cannot be deleted. Public Gateway: r006-4d6d4828-b95b-471b-85f8-a84147e596fb
  Error target name: id, type: parameter
  Error target value: r006-17e4a0d2-81bc-407b-a74d-00a53fa314bd
  Trace ID: 2f889eac-ff21-4387-8ba8-c3889e67ca32
  
  Delete incomplete, full delete can be done when vpc has no cluster/instances/subnets etc... in it.
  A cluster/cloud resource can only be deleted once it's attached resources are released.
  Please try again, after the cluster(s) delete is complete.
  ```

ScrapCodes commented:
Hi @yhwang, this is the question I had. It takes about 45+ mins to fully delete a cluster, before subnets can be deleted. Do you think the user experience will be great, if the script hangs for 45+ mins waiting for cluster delete? 

Currently, one can wait for some time, and re run the script to delete the subnet/pubgw/vpc etc...

yhwang commented:
understood. maybe having an option to wait for the deletion would be good enough. default value of it could be "off" and users can explicitly turn it "on" to wait for the cluster being deleted

Tomcli commented:
@yhwang is this PR looks good to you?

yhwang commented:
@Tomcli like my previous comment, @ScrapCodes can you add an option to wait for the cluster clean up?

ScrapCodes commented:
@yhwang and @Tomcli I am still working on this one.

ScrapCodes commented:
@yhwang Thank you for the review. Please take a look again !

ScrapCodes commented:
@yhwang Gentle ping !

yhwang commented:
@ScrapCodes sorry for the delay. I was verifying the 1.3 release last week. let me try this PR today and update you afterwards.

yhwang commented:
- `./deploy-ibm-vpc.sh --cluster-name="yhwang-script-pr" --vpc-name="yhwang-script-pr-vpc" --deploy-kubeflow=true`
  After the command finished, I hit this error:
  ```
  You can now execute 'kubectl' commands against your cluster. For example, run 'kubectl get nodes'.
  If you are accessing the cluster for the first time, 'kubectl' commands might fail for a few seconds while RBAC synchronizes.
  The cluster: yhwang-script-pr in zone: us-south-3, is ready.
  Error from server (Forbidden): nodes is forbidden: User "IAM#yh.wang@ibm.com" cannot list resource "nodes" in API group "" at the cluster scope
  ```
  so I guess you may need to retries the `kubectl` command several times for the first access. Or just let users run the
  command again. I ran the command again to finish the kubeflow deployment.
- The ingress subdomain creation seems time consuming. But only solution is to wait.
- Then I used this command to delete the cluster:
  `./deploy-ibm-vpc.sh --delete-cluster="full-sync" --config-file=~/.iks/yhwang-script-pr-vpc-yhwang-script-pr-iks-conf.sh --vpc-name=yhwang-script-pr-vpc --cluster-name=yhwang-script-pr`
  One question is: I assigned the `--config-file` in the command file, why the `--vpc-name` and `--cluster-name` are still mandatory?
- after the `delete-cluster` completed, it showed me this:
  ```
  SUBNET_ID and GATEWAY_ID is not defined skipping...
  Delete incomplete, full delete can be done when vpc has no cluster/instances/subnets etc... in it.
  A cluster/cloud resource can only be deleted once it's attached resources are released.
  Please check if VPC has other resources in it.
  ```
- ran the delete command again, still got the same error.

Tomcli commented:
thanks @ScrapCodes @yhwang. I don't want you guys spend too much time on this PR. For the delete part we can have some manual instructions since it's not very critical. For this script I'm mainly looking for automating the cluster creation process.

yhwang commented:
@Tomcli if that's the case, then the cluster creation part is okay to me. If there is any error, user just need to re-run the command again. For me, it took me twice.

ScrapCodes commented:
Hi @yhwang 
Actually, I did not see those errors earlier. May be they happen once in a while.
So, I have some delays between steps, that should actually resolve for most cases. In those rare cases, we have already described in the Readme that a user can retry the same command again.

To test out after adding the delays between steps, I ran your test case as follows:
```
deploy-ibm-vpc.sh --config-file=`pwd`/.iks/cc.sh --delete-cluster=full-sync
Environment variables for cluster ...
# This is a config file for the cluster: xyz in VPC: yhwang-script-pr-vpc
export SUBNET_ID=0737-9bfe9e35-0fa0-4d23-8740-92acc4fbf406
export VPC_ID=r006-395e0527-c42e-44d9-a5ea-c8c7a265ed03
export VPC_NAME=yhwang-script-pr-vpc
export GATEWAY_ID=r006-8318905e-7fe9-4412-ba5c-16e24cd4c15b
export KUBERNETES_VERSION=1.18
export CLUSTER_ZONE=us-south-3
export CLUSTER_NAME=xyz
export WORKER_NODE_FLAVOR=bx2.4x16
export WORKER_COUNT=2
ibmcloud is already installed,
make sure you are logged-in by either `ibmcloud login `
 or ` ibmcloud login --sso`.
done.
Cluster delete failed.
Waiting for the cluster to be deleted, this usually takes about 45 mins.
Subnet 0737-9bfe9e35-0fa0-4d23-8740-92acc4fbf406 is deleted.
Public gateway r006-8318905e-7fe9-4412-ba5c-16e24cd4c15b is deleted.
vpc r006-395e0527-c42e-44d9-a5ea-c8c7a265ed03 is deleted.
VPC r006-395e0527-c42e-44d9-a5ea-c8c7a265ed03 is deleted. Complete delete successful.
```

Thanks,

yhwang commented:
/lgtm

Tomcli commented:
/lgtm
/approve
